src/app/campaigns/campaign-edit/campaign-edit.component.ts

```diff
@@ -23,6 +23,7 @@ export class CampaignEditComponent {
   campaign$: Observable<WithStatus<CampaignWithInfo>>
 
   updateForm: FormGroup
+  newsUrls: FormArray
```

Comment:
It would be better not to have this `newsUrls` in component state, in my opinion. you can pass a form control in every function you need (like addNewsUrl, removeNewsUrl) as you are iterating over them.

Reply:
I tried rearranging this, but in the end I find this to be the cleanest way since we need `newsUrls` form array in couple of places. Also, all of the examples I saw put it into the component state directly.

Reply:
> Also, all of the examples I saw put it into the component state directly.

It is generally a bad practice to hold a subset of the same state in multiple state variables. That's why we are rarely using plain variables in components and use observables to chain them if needed. 
When you see the examples, be concerned about the best practices in them.. They don't give a f**k about it, just to show the API. 
If you ask me, it is much much better to pass the parameter to the function that needs it instead of using a shared variable. it's easier to "test" and mentally accept what it does and what is the variable state in a particular moment. 
however, if you don't agree, you can leave it like this.

Reply:
Not having form array in a state variable is kind of awkward since you always have to do a cast: `this.updateForm.get('newsUrls') as FormArray`. If we want to move it from the state, then we can move that into a function. Passing that form array as an argument to `newsUrlsControls()`, `addNewsUrl()` and `removeNewsUrl(index)` makes sense if we rename them to something more generic, e.g. `formArrayControls(formArray)`, `addFormArrayControl(formArray)` and `removeFormArrayControl(formArray)`.

Reply:
I agree. After some testing and trying to make it prettier, this is "OK" way for now, but we'll try to make it easier to maintain. Consider having 5 form arrays on several edit pages in the app, we would need to duplicate all this silly logic around. I'm not happy with this, but we'll fix it in the future! :)